templates/install_puppet.ps1.erb

```diff
@@ -0,0 +1,262 @@
+  # to actually throw [System.Management.Automation.ActionPreferenceStopException] rather than the actual
+  # exceptions. Once inside the catch block, the value of $_ is correctly set as the actual exception
+  } catch {
+    if ($_.Exception.GetType() -eq [System.TimeoutException]) {
+      Write-Log "ERROR: timed out waiting for puppet run to finish!"
+      throw $_
+    } else {
+      Write-Log "ERROR: failed while trying to wait for puppet process with $($_.Exception.GetType())"
+      throw $_
+    }
+  }
+  # We *must* shutdown all puppet-agent services for the MSI installation to correctly
+  # work without requiring a restart.
+  foreach($service in $service_names) {
+    $serv_exists = Get-Service -Name $service -ErrorAction SilentlyContinue
+    if ($serv_exists) {
+      Write-Log "Stopping $($service) before upgrade"
+      Stop-Service $service
+    }
+  }
+  if ($UseLockedFilesWorkaround) {
+    $temp_puppetres = Move-PuppetresDLL
+  }
+  $msi_arguments = "/qn /norestart /i `"$Source`" /l*vx+ `"$Logfile`""
+  $msi_arguments += " PUPPET_AGENT_ENVIRONMENT=`"$((puppet.bat config print --section agent environment) -replace '\s','')`""
+  if ($InstallDir) {
+    $msi_arguments += " INSTALLDIR=`"$InstallDir`""
+  }
+  if ($PuppetMaster) {
+    $msi_arguments += " PUPPET_MASTER_SERVER=`"$PuppetMaster`""
+  }
+  if ($PuppetStartType) {
+    $msi_arguments += " PUPPET_AGENT_STARTUP_MODE=`"$PuppetStartType`""
+  }
+  $msi_arguments += " $InstallArgs"
+  Write-Log "Beginning MSI installation with Arguments: $msi_arguments"
+  Write-Log "****************************** Begin msiexec.exe output ******************************"
+  $startInfo = New-Object System.Diagnostics.ProcessStartInfo('msiexec.exe', $msi_arguments)
+  $startInfo.UseShellExecute = $false
+  $startInfo.CreateNoWindow = $true
+  $invocationId = [Guid]::NewGuid().ToString()
+  $process = New-Object System.Diagnostics.Process
+  $process.StartInfo = $startInfo
+  $process.EnableRaisingEvents = $true
+  $exitedEvent = Register-ObjectEvent -InputObject $process -EventName 'Exited' -SourceIdentifier $invocationId
+  $process.Start() | Out-Null
+  # park current thread until the PS event is signaled upon process exit
+  # OR the timeout has elapsed
+  $waitResult = Wait-Event -SourceIdentifier $invocationId
+  Write-Log "****************************** End msiexec.exe output ******************************"
+  if (($process.ExitCode -eq 3010) -or ($process.ExitCode -eq 1641) ){
+    Write-Log "WARNING: msiexec.exe returned $($msi_exec.ExitCode) and has flagged the system for restart!!!"
+  } elseif ($process.ExitCode -ne 0){
+    Write-Log "ERROR: msiexec.exe installation failed!!! Return code $($msi_exec.ExitCode)"
+    throw
+  }
+} catch {
+  Write-Log "ERROR: $_"
+  if ($UseLockedFilesWorkaround) {
+    Reset-PuppetresDLL $temp_puppetres
+  }
+} finally {
+  Unlock-Installation $install_pid_lock

```

Comment:
I think you also want to call `Unregister-Event` to perform cleanup... though exiting PowerShell altogether may do this for you